### PR TITLE
[ᚬrc/v0.12.0] fix(network): Avoid dial too often

### PR DESCRIPTION
**Pull request by TheWaWaR**

### CODE REVIEW

File: network/src/network.rs [hunk 1]
```diff
@@ -337,26 +339,91 @@ impl NetworkState {
...
+        if self.listened_addresses.read().contains_key(&addr) {
+            trace!(target: "network", "Do not dial listened address(self): {:?}, {}", peer_id, addr);
+            return false;
+        }
+
+        if self
+            .failed_dials
+            .lock()
+            .get_refresh(peer_id)
+            .map(|last_dial| last_dial.elapsed() <= Duration::from_secs(300))
```
Comment (doitian) [inline network/src/network.rs]: Define a constant for it.

File: network/src/network.rs [hunk 2]
```diff
@@ -337,26 +343,90 @@ impl NetworkState {
...
+        }
+
+        if self
+            .failed_dials
+            .lock()
+            .get_refresh(peer_id)
+            .map(|last_dial| last_dial.elapsed() <= DIAL_FAILED_WAIT)
+            .unwrap_or(false)
+        {
+            trace!(target: "network", "Do not dial address fialed less than 5 minutes: {:?}, {}", peer_id, addr);
```
Comment (doitian) [inline network/src/network.rs]: Use `DIAL_FAILED_WAIT` here to avoid inconsistent message.

File: network/src/network.rs [hunk 3]
```diff
@@ -337,26 +343,90 @@ impl NetworkState {
...
+            trace!(target: "network", "Do not dial peer in registry: {:?}, {}", peer_id, addr);
+            return false;
+        }
+
+        if let Some(dial_started) = self.dialing_addrs.read().get(peer_id) {
+            trace!(target: "network", "Do not repeat send dial command to network service: {:?}, {}", peer_id, addr);
+            if dial_started.elapsed() > DIAL_HANG_TIMEOUT {
+                error!(
+                    target: "network",
+                    "Dialing {:?}, {:?} for more than 5 minutes, something is wrong in network service",
```
Comment (doitian) [inline network/src/network.rs]: Use `DIAL_HANG_TIMEOUT` here to avoid an inconsistent message.

File: network/src/services/outbound_peer.rs [hunk 4]
```diff
@@ -70,7 +55,7 @@ impl OutboundPeerService {
         let p2p_control = self.p2p_control.clone();
         for (peer_id, addr) in peers
             .into_iter()
-            .filter(|(peer_id, _addr)| self.network_state.local_peer_id() != peer_id)
+            .filter(|(peer_id, addr)| self.network_state.can_dial(peer_id, addr))
```
Comment (jjyr) [inline network/src/services/outbound_peer.rs]: duplicate filter? can_dial is already called in dial method

File: network/src/network.rs [hunk 5]
```diff
@@ -522,6 +588,15 @@ impl ServiceHandle for EventHandler {
...
                     session_context.address,
                 );
+                let peer_id = session_context
+                    .remote_pubkey
+                    .as_ref()
+                    .map(PublicKey::peer_id)
+                    .expect("Secio must enabled");
+
+                self.network_state.with_peer_registry_mut(|reg| {
+                    reg.remove_feeler(&peer_id);
```
Comment (jjyr) [inline network/src/network.rs]: missing disconnect?

File: network/src/network.rs [hunk 6]
```diff
@@ -337,26 +343,90 @@ impl NetworkState {
...
+                    addr,
+                );
+            }
+            return false;
+        }
+
+        true
+    }
+
+    pub(crate) fn dial_fialed(&self, peer_id: PeerId) {
```
Comment (jjyr) [inline network/src/network.rs]: failed

File: network/src/network.rs [hunk 7]
```diff
@@ -337,26 +343,90 @@ impl NetworkState {
...
+            reg.get_key_by_peer_id(peer_id).is_some() || reg.is_feeler(peer_id)
+        });
+        if peer_in_registry {
+            trace!(target: "network", "Do not dial peer in registry: {:?}, {}", peer_id, addr);
+            return false;
+        }
+
+        if let Some(dial_started) = self.dialing_addrs.read().get(peer_id) {
+            trace!(target: "network", "Do not repeat send dial command to network service: {:?}, {}", peer_id, addr);
+            if dial_started.elapsed() > DIAL_HANG_TIMEOUT {
```
Comment (jjyr) [inline network/src/network.rs]: consider to send this message to sentry?

File: network/src/services/outbound_peer.rs (hunk 4 above)
Comment (TheWaWaR) [reply network/src/services/outbound_peer.rs]: dial may called elsewhere

File: network/src/network.rs (hunk 5 above)
Comment (TheWaWaR) [reply network/src/network.rs]: IT IS a disconnected session.

File: network/src/network.rs (hunk 7 above)
Comment (TheWaWaR) [reply network/src/network.rs]: This will be called frequently, we will noticed as soon as possiable

**Pull request merged**